src/libsystemd-network/dhcp-option.c

```diff
@@ -356,10 +349,168 @@ static int parse_options(const uint8_t options[], size_t buflen, uint8_t *overlo
+        _cleanup_merge_option_node_ MergeOptionNode *next = NULL;
+
+        while (offset < buflen) {
+                const uint8_t *data = &buf[offset];
+                uint8_t code = buf[offset++];
+
+                switch (code) {
+                case SD_DHCP_OPTION_PAD:
+                        continue;
+                case SD_DHCP_OPTION_END:
+                        return 0;
+                }
+
+                if (buflen < offset + 1)
+                        return -ENOBUFS;
+
+                uint8_t len = buf[offset++];
+
+                if (buflen < offset + len)
+                        return -EINVAL;
+
+                switch (code) {
+                case SD_DHCP_OPTION_OVERLOAD:
+                        if (len != 1)
+                                return -EINVAL;
+                        if (overload)
+                                *overload = buf[offset];
+                        offset++;
+                        continue;
+                }
+
+                next = new(MergeOptionNode, 1);
+                if (!next)
+                        return -ENOMEM;
+                *next = (MergeOptionNode){.data = data};
+
+                MergeOptionNode *head = ordered_hashmap_get(opts, UINT_TO_PTR(code));
+                if (head)
+                        LIST_APPEND(options, head, TAKE_PTR(next));
+                else {
+                        r = ordered_hashmap_put(opts, UINT_TO_PTR(code), next);
+                        if (r < 0)
+                                return r;
+                        TAKE_PTR(next);
+                }
+
+                offset += len;
+                *tot_len += len;
+        }
+
+        return 0;
+}
+
+/* RFC3396 specifies how DHCP options longer than 255 bytes should be handled.
+ *
+ * Option is to be repeated multiple times and treated as one, long, option. However, the cut can
+ * happen at byte boundaries, and clients MUST assume no semantic meaning whatsoever from this cut.
+ * In practice, this means the cut can happen anywhere in the middle of a semantic unit of the
+ * option (e.g. DHCP option 121 (Classless Static Routes) can be split in the middle of a route).
+ *
+ * With the current architecture of stateless callbacks that take a whole chunk of memory, it
+ * becomes hard to retain data between two consecutive callbacks without major changes to the
+ * architecture.
+ *
+ * This function will merge DHCP options by concatenating options of the same type, producing a new
+ * TLV buffer, but using u16 len fields instead of u8. This works due to a DHCP message size having
+ * an inherent upper bound, lower than u16::MAX.
+ *
+ * Basically, this produces the aggregated option buffer the RFC talks about.
+ */
+static int dhcp_option_merge(const DHCPMessage *message, size_t buflen, uint8_t **ret_merged) {
+        _cleanup_ordered_hashmap_free_ OrderedHashmap *opts;
+        uint8_t *aggregate, overload = 0;
+        size_t tot_len = 0;
+        int r = 0;
+
```

Comment:
we generally put `assert()` lines at the top of functions that check relevant parameters for non-NULL-ness.

Reply:
Done.